### PR TITLE
do not leak api-server cluster ip from seed to shoot

### PR DESCRIPTION
**Pull request by domdom82**
**How to categorize this PR?**

/area networking
/kind cleanup

**What this PR does / why we need it**:

The cluster IP address of the kube-apiserver service on the seed should not leak into the shoot data plane. The IP address is fairly hard to predict and may overlap with some IP address with which the shoot data plane wants to communicate. Therefore, a different IP address should be used.

The PR fixes this issue by mapping the seed cluster IP to an otherwise unusable reserved range of `240.0.0.0/4`. The three last bytes of the cluster ip get prefixed by `240.` making the original IP usable in the shoot again. At the same time, shoots are forbidden from using the reserved IP range anywhere.

**Special notes for your reviewer**:
- Please point to missing docs that need to be updated

**Release note**:

```breaking operator
This change applies to IPv4 clusters only.
Gardener uses the CIDR range of 240.0.0.0/8 which is reserved as per [IANA db](https://www.iana.org/assignments/ipv4-address-space/ipv4-address-space.xhtml) to map the cluster ip of the kubernetes api-server in the seed to a different network range before exposing it to the shoot in the `kubernetes` service. This frees up address space in the shoot and removes potential clashes with shoot workload ips.

Seed operators need to check if any of the following properties collide with the 240.0.0.0/8 range:

    spec:
      networks:
        pods: < check here >
        nodes: < check here >
        services: < check here >
        shootDefaults:
          pods: < check here >
          nodes: < check here >
          services: < check here >

```

### CODE REVIEW

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 1]
```diff
@@ -781,7 +781,10 @@ const (
 	DefaultVPNRange = "192.168.123.0/24"
 	// DefaultVPNRangeV6 is the default IPv6 network range for the VPN between seed and shoot cluster.
 	DefaultVPNRangeV6 = "fd8f:6d53:b97a:1::/96"
-
+	// ReservedRangeApiServerProxy is the IPv4 network range for the "kubernetes" service used by apiserver-proxy
+	ReservedRangeApiServerProxy = "240.0.0.0/4"
+	// SeedServiceRangeMaxSizeBits is the maximum size of the service network for a seed. Must fit into ReservedRangeApiServerProxy.
+	SeedServiceRangeMaxSizeBits = 8
```
Comment (ScheererJ) [inline pkg/apis/core/v1beta1/constants/types_constants.go]: Why specify a `/4` when we only use `/8`? This feels inconsistent.

We can either make the chance of the service range being too large smaller by actually using the `/4` and allowing service ranges up to `/4` or we actually stick with `/8`, which is already extremely large, and makes the bit handling a bit easier, but then we should adapt the range as well.

Comment (ScheererJ) [reply pkg/apis/core/v1beta1/constants/types_constants.go]: Please add an empty line afterwards, though.

File: pkg/apis/core/validation/seed.go [hunk 2]
```diff
@@ -256,7 +256,12 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
 		networks = append(networks, cidrvalidation.NewCIDR(seedNetworks.Pods, fldPath.Child("pods")))
 	}
 	if !inTemplate || len(seedNetworks.Services) > 0 {
-		networks = append(networks, cidrvalidation.NewCIDR(seedNetworks.Services, fldPath.Child("services")))
+		servicesNw := cidrvalidation.NewCIDR(seedNetworks.Services, fldPath.Child("services"))
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed.go]:
```suggestion
		services := cidrvalidation.NewCIDR(seedNetworks.Services, fldPath.Child("services"))
```

File: pkg/apis/core/validation/seed.go [hunk 3]
```diff
@@ -277,6 +282,8 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
 	}
 	allErrs = append(allErrs, cidrvalidation.ValidateCIDROverlap(networks, false)...)
 
+	apiServerProxyRange := cidrvalidation.NewCIDR(v1beta1constants.ReservedRangeApiServerProxy, field.NewPath(""))
+	allErrs = append(allErrs, apiServerProxyRange.ValidateNotOverlap(networks...)...)
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed.go]: While it is unlikely someone might have created a seed cluster overlapping with this range. How do we handle this?

File: pkg/apis/core/validation/seed_test.go [hunk 4]
```diff
@@ -535,6 +535,30 @@ var _ = Describe("Seed Validation Tests", func() {
 						"Detail": Equal(`must not overlap with "[]" ("192.168.123.0/24")`),
 					}))
 				})
+				It("should forbid Seed with overlap to reserved apiserver-proxy range", func() {
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed_test.go]:
```suggestion

				It("should forbid Seed with overlap to reserved apiserver-proxy range", func() {
```

File: pkg/apis/core/validation/seed_test.go [hunk 5]
```diff
@@ -535,6 +535,30 @@ var _ = Describe("Seed Validation Tests", func() {
...
+
+					errorList := ValidateSeed(seed)
+
+					Expect(errorList).To(ConsistOfFields(Fields{
+						"Type":   Equal(field.ErrorTypeInvalid),
+						"Field":  Equal("spec.networks.pods"),
+						"Detail": Equal(`must not overlap with "[]" ("240.0.0.0/4")`),
+					}))
+				})
+				It("should forbid Seed with too large service range", func() {
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed_test.go:552]:
```suggestion

				It("should forbid Seed with too large service range", func() {
```

File: pkg/apis/core/validation/seed_test.go [hunk 6]
```diff
@@ -616,6 +640,17 @@ var _ = Describe("Seed Validation Tests", func() {
 						"Detail": ContainSubstring("must be a valid IPv6 address"),
 					}))
 				})
+				It("should allow Seed with very large service range because the range limit only applies to ipv4", func() {
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed_test.go:646]:
```suggestion

				It("should allow Seed with very large service range because the range limit only applies to ipv4", func() {
```

File: pkg/apis/core/validation/shoot.go [hunk 7]
```diff
@@ -932,6 +932,7 @@ func validateNetworking(networking *core.Networking, workerless bool, fldPath *f
 	}
 
 	primaryIPFamily := helper.DeterminePrimaryIPFamily(networking.IPFamilies)
+	apiServerProxyRange := cidrvalidation.NewCIDR(v1beta1constants.ReservedRangeApiServerProxy, field.NewPath(""))
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot.go]: Please check how the overlap with the VPN ranges is done and do the same for this instead of checking here.

The question is how to deal with existing clusters. Those should not break completely.

File: pkg/utils/validation/cidr/cidr.go [hunk 8]
```diff
@@ -41,6 +41,10 @@ type CIDR interface {
 	LastIPInRange() net.IP
 	// ValidateOverlap returns errors if the subnets do not overlap with CIDR.
 	ValidateOverlap(subsets ...CIDR) field.ErrorList
+	// ValidateMaxSize returns errors if the subnet is larger than the given bits. e.g. /15 is larger than /16
+	ValidateMaxSize(bits int) field.ErrorList
+	// ValidateMinSize returns errors if the subnet is smaller than the given bits. e.g. /16 is smaller than /15
+	ValidateMinSize(bits int) field.ErrorList
```
Comment (ScheererJ) [inline pkg/utils/validation/cidr/cidr.go]: I do not see that we use this function. Hence, please remove it. We can add it once we need it.

**Review by ScheererJ — COMMENTED** [review]
Thanks for addressing this topic.

Could you please clarify in the PR description that this is only for IPv4 because of potential overlap in the private address space? Furthermore, please add a release note.

I am not aware of having the VPN range limitation in the documentation. Hence, I am not sure where to document this one.

File: pkg/apis/core/v1beta1/constants/types_constants.go (hunk 1 above)
Comment (domdom82) [reply pkg/apis/core/v1beta1/constants/types_constants.go]: I wanted to keep the range as it was in the RFC but you're right if we're not going to use it we might as well forego the 4 bits in the range and not just the code. It will also allow us to remove the `SeedServiceRangeMaxSizeBits` property because now we can just use the size of `ReservedRangeApiServerProxy` instead. I'll address it.

File: pkg/utils/validation/cidr/cidr_test.go [hunk 9]
```diff
@@ -274,6 +274,63 @@ var _ = Describe("cidr", func() {
...
+
+		Describe("ValidateMaxSize", func() {
+			It("should return no errors if cidr is same size than limit", func() {
+				goodPath := field.NewPath("good")
+				goodCIDR := "11.12.0.0/16"
+				good := NewCIDR(goodCIDR, goodPath)
+
+				Expect(good.ValidateMaxSize(16)).To(BeEmpty())
+			})
+			It("should return no errors if cidr is smaller than limit", func() {
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go:287]: Add a newline between `It`s

File: pkg/utils/validation/cidr/cidr_test.go [hunk 10]
```diff
@@ -274,6 +274,63 @@ var _ = Describe("cidr", func() {
...
+
+				Expect(bad.ValidateMaxSize(16)).To(ConsistOfFields(Fields{
+					"Type":     Equal(field.ErrorTypeInvalid),
+					"Field":    Equal(badPath.String()),
+					"BadValue": Equal(badCIDR),
+					"Detail":   Equal(`cannot be larger than /16`),
+				}))
+			})
+		})
+		Describe("ValidateMinSize", func() {
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go]:
```suggestion

		Describe("ValidateMinSize", func() {
```

File: pkg/utils/validation/cidr/cidr_test.go [hunk 11]
```diff
@@ -520,5 +577,61 @@ var _ = Describe("cidr", func() {
 				Expect(cdr.ValidateOverlap(other)).To(BeEmpty())
 			})
 		})
+		Describe("ValidateMaxSize", func() {
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go:587]:
```suggestion

		Describe("ValidateMaxSize", func() {
```

File: pkg/utils/validation/cidr/cidr_test.go [hunk 12]
```diff
@@ -520,5 +577,61 @@ var _ = Describe("cidr", func() {
...
+
+				Expect(bad.ValidateMaxSize(32)).To(ConsistOfFields(Fields{
+					"Type":     Equal(field.ErrorTypeInvalid),
+					"Field":    Equal(badPath.String()),
+					"BadValue": Equal(badCIDR),
+					"Detail":   Equal(`cannot be larger than /32`),
+				}))
+			})
+		})
+		Describe("ValidateMinSize", func() {
```
Comment (rfranzke) [inline pkg/utils/validation/cidr/cidr_test.go]:
```suggestion

		Describe("ValidateMinSize", func() {
```

File: pkg/apis/core/validation/seed.go (hunk 3 above)
Comment (domdom82) [reply pkg/apis/core/validation/seed.go]: Release notes? There are more reserved ranges we don't check that could create all kinds of problems, e.g. multicast range. IMO gardener should have checked those from the beginning and forbid them.

File: pkg/utils/validation/cidr/cidr.go (hunk 8 above)
Comment (domdom82) [reply pkg/utils/validation/cidr/cidr.go]: I added it for completeness but fair enough.

File: pkg/apis/core/validation/shoot.go (hunk 7 above)
Comment (domdom82) [reply pkg/apis/core/validation/shoot.go]: moved to cidr validation where vpn ranges are checked also

File: pkg/utils/validation/cidr/cidr.go (hunk 8 above)
Comment (domdom82) [reply pkg/utils/validation/cidr/cidr.go]: function removed

**Comment by domdom82** [comment]
Findings during shoot upgrade:
- During reconcile I observed a flapping between the old cluster-ip and new mapped ip on the `kubernetes` service endpoint until it finally settled on the new ip:
```
k get endpoints kubernetes --watch
NAME         ENDPOINTS            AGE
kubernetes   100.107.17.226:443   61m
kubernetes   100.107.17.226:443,240.107.17.226:443   61m
kubernetes   100.107.17.226:443                      61m
kubernetes   100.107.17.226:443,240.107.17.226:443   61m
kubernetes   240.107.17.226:443                      62m
```
- As expected, the node now has both the old and the new ip on loopback:
```
ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 100.107.17.226/32 brd 100.107.17.226 scope host lo
       valid_lft forever preferred_lft forever
    inet 240.107.17.226/32 brd 240.107.17.226 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute
       valid_lft forever preferred_lft forever
```

- Connecting to the `kubernetes` service yields the cluster-ip `100.112.0.1` as before, now pointing to the new mapped ip `240.107.17.226` under the hood:
```
root@test:/# curl -vk https://kubernetes.default.svc.cluster.local
* Host kubernetes.default.svc.cluster.local:443 was resolved.
* IPv6: (none)
* IPv4: 100.112.0.1
*   Trying 100.112.0.1:443...
* Connected to kubernetes.default.svc.cluster.local (100.112.0.1) port 443
(...)

> GET / HTTP/2
> Host: kubernetes.default.svc.cluster.local
> User-Agent: curl/8.5.0
> Accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* received GOAWAY, error=0, last_stream=1
< HTTP/2 401
< audit-id: fc46c5e5-8390-4d82-a368-969c0cb86ba6
< cache-control: no-cache, private
< content-type: application/json
< content-length: 157
< date: Tue, 10 Dec 2024 12:05:26 GMT
<
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
* Closing connection
```

works fine as before.

- The old seed cluster-ip is no longer used, envoy is listening only on `240.107.17.226` now:
```
# netstat -antp | grep LISTEN
tcp        0      0 127.0.0.1:36239         0.0.0.0:*               LISTEN      1655/containerd
tcp        0      0 169.254.20.10:8099      0.0.0.0:*               LISTEN      40133/node-cache
tcp        0      0 100.112.0.10:53         0.0.0.0:*               LISTEN      40133/node-cache

tcp        0      0 240.107.17.226:443      0.0.0.0:*               LISTEN      33017/envoy
tcp        0      0 240.107.17.226:443      0.0.0.0:*               LISTEN      33017/envoy

tcp        0      0 127.0.0.54:53           0.0.0.0:*               LISTEN      695/systemd-resolve
tcp        0      0 127.0.0.1:10248         0.0.0.0:*               LISTEN      3126/kubelet
tcp        0      0 169.254.20.10:53        0.0.0.0:*               LISTEN      40133/node-cache
tcp        0      0 0.0.0.0:16910           0.0.0.0:*               LISTEN      33017/envoy
tcp        0      0 0.0.0.0:16910           0.0.0.0:*               LISTEN      33017/envoy
tcp        0      0 127.0.0.1:9099          0.0.0.0:*               LISTEN      6837/calico-node
tcp        0      0 127.0.0.1:9098          0.0.0.0:*               LISTEN      6802/calico-typha
tcp        0      0 127.0.0.1:5000          0.0.0.0:*               LISTEN      33332/registry
tcp        0      0 0.0.0.0:111             0.0.0.0:*               LISTEN      1/systemd
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1325/sshd: /usr/sbi
tcp        0      0 127.0.0.53:53           0.0.0.0:*               LISTEN      695/systemd-resolve
```

- As expected, deleting the old ip address has no effect:
```
# ip address delete 100.107.17.226/32 dev lo

(...)

curl -vk https://kubernetes.default.svc.cluster.local
* Host kubernetes.default.svc.cluster.local:443 was resolved.
* IPv6: (none)
* IPv4: 100.112.0.1
*   Trying 100.112.0.1:443...
* Connected to kubernetes.default.svc.cluster.local (100.112.0.1) port 443

(...)

> GET / HTTP/2
> Host: kubernetes.default.svc.cluster.local
> User-Agent: curl/8.5.0
> Accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* received GOAWAY, error=0, last_stream=1
< HTTP/2 401
< audit-id: 16c456c8-f137-470c-905d-ebe3385a4c8e
< cache-control: no-cache, private
< content-type: application/json
< content-length: 157
< date: Tue, 10 Dec 2024 12:17:46 GMT
<
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
* Closing connection
```

So, in conclusion:
- yes, the old ip remains on the node until the node is rolled
- but it disappears from the service and is no longer used anywhere
- removing the ip has a cosmetic effect at best

File: docs/usage/networking/shoot_networking.md [hunk 13]
```diff
@@ -85,3 +85,17 @@ Number of IPs per podCIDRs: 128
...
 
 With the configuration above, a Shoot cluster can at most have **32 nodes** which are ready to run workload in the Pod network.
+
+## Reserved Networks
+
+Some network ranges are reserved for specific use-cases in the communication between seeds and shoots.
+
+| IPv  | CIDR                  | Name               | Purpose                                                                    |
+|------|-----------------------|--------------------|----------------------------------------------------------------------------|
+| IPv4 | 192.168.123.0/24      | Default VPN Range  | Used for communication between seed API server and shoot resources via VPN |
```
Comment (ScheererJ) [inline docs/usage/networking/shoot_networking.md]: Maybe we should add that this will be removed from the list once feature gate `NewVPN` is graduated.

Apart from that, it is good to have this section in the documentation.

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 14]
```diff
@@ -781,6 +781,8 @@ const (
 	DefaultVPNRange = "192.168.123.0/24"
 	// DefaultVPNRangeV6 is the default IPv6 network range for the VPN between seed and shoot cluster.
 	DefaultVPNRangeV6 = "fd8f:6d53:b97a:1::/96"
+	// ReservedSeedServiceRange is the IPv4 network range for the "kubernetes" service used by apiserver-proxy
+	ReservedSeedServiceRange = "240.0.0.0/8"
```
Comment (ScheererJ) [inline pkg/apis/core/v1beta1/constants/types_constants.go]: I am a bit unhappy with the naming. This range will only show up in shoot clusters, but its name suggests it is connected to the seed service range. Maybe something like `ReservedSeedServiceReplacementRange` fits better, but I open to other names.

File: pkg/apis/core/validation/seed.go [hunk 15]
```diff
@@ -248,15 +248,22 @@ func validateSeedNetworks(seedNetworks core.SeedNetworks, fldPath *field.Path, i
...
 
 	if !inTemplate || len(seedNetworks.Pods) > 0 {
 		networks = append(networks, cidrvalidation.NewCIDR(seedNetworks.Pods, fldPath.Child("pods")))
 	}
 	if !inTemplate || len(seedNetworks.Services) > 0 {
-		networks = append(networks, cidrvalidation.NewCIDR(seedNetworks.Services, fldPath.Child("services")))
+		services := cidrvalidation.NewCIDR(seedNetworks.Services, fldPath.Child("services"))
+		networks = append(networks, services)
+		// Service range must not be larger than /8 for ipv4
+		if len(services.ValidateIPFamily(cidrvalidation.IPFamilyIPv4)) == 0 {
```
Comment (ScheererJ) [inline pkg/apis/core/validation/seed.go]: Nit: Personally, I find the `if` condition hard to grasp as it implies the reader knows that the validate function returns a slice of errors and an empty list implies that the service range uses IPv4. Maybe a simple helper function could be less confusing.

File: pkg/gardenlet/operation/botanist/kubeapiserverexposure.go [hunk 16]
```diff
@@ -68,11 +68,21 @@ func (b *Botanist) DeployKubeAPIServerSNI(ctx context.Context) error {
...
+	if clusterIPv4 != nil {
+		if b.Shoot.GetInfo().Spec.Networking.IPFamilies[0] == v1beta1.IPFamilyIPv6 {
+			// "64:ff9b:1::" is a well known prefix for address translation for use
+			// in local networks.
+			b.APIServerClusterIP = "64:ff9b:1::" + clusterIP
+		} else {
+			// prevent leakage of real cluster ip to shoot. we use the reserved range 240.0.0.0/8 as prefix instead.
+			prefixIp, _, _ := net.ParseCIDR(v1beta1constants.ReservedSeedServiceRange)
+			prefix := prefixIp.To4()
+			b.APIServerClusterIP = net.IPv4(prefix[0], clusterIPv4[1], clusterIPv4[2], clusterIPv4[3]).String()
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/kubeapiserverexposure.go]: Let's add an example in the comment, e.g.

```
// cluster ip in seed:  192.168.102.23
// ip in shoot:         240.168.102.23
```

File: pkg/utils/validation/cidr/disjoint.go [hunk 17]
```diff
@@ -65,6 +65,10 @@ func validateOverlapWithSeed(fldPath *field.Path, shootNetwork []string, network
 		if NetworksIntersect(v1beta1constants.DefaultVPNRangeV6, network) {
 			allErrs = append(allErrs, field.Invalid(fldPath, network, fmt.Sprintf("shoot %s network intersects with default vpn network (%s)", networkType, v1beta1constants.DefaultVPNRangeV6)))
 		}
+
+		if NetworksIntersect(v1beta1constants.ReservedSeedServiceRange, network) {
+			allErrs = append(allErrs, field.Invalid(fldPath, network, fmt.Sprintf("shoot %s network intersects with reserved seed service range (%s)", networkType, v1beta1constants.ReservedSeedServiceRange)))
```
Comment (ScheererJ) [inline pkg/utils/validation/cidr/disjoint.go]: Maybe we should call it "reserved kube-apiserver service range" instead.

**Review by ScheererJ — COMMENTED** [review]
Thanks for adding the documentation. It is good to have something to point to.
I have a few additional comments.

File: pkg/apis/core/v1beta1/constants/types_constants.go (hunk 14 above)
Comment (domdom82) [reply pkg/apis/core/v1beta1/constants/types_constants.go]: `ReservedSeedServiceMappingRange`?

Comment (domdom82) [reply pkg/apis/core/v1beta1/constants/types_constants.go]: `ReservedKubeApiServerMappingRange`?

**Review by ScheererJ — APPROVED** [review]
/lgtm
/approve
